core/src/main/java/org/apache/iceberg/TableMetadataParser.java

```diff
@@ -253,8 +254,42 @@ public static TableMetadata read(FileIO io, InputFile file) {
     }
   }
 
-  @SuppressWarnings("checkstyle:CyclomaticComplexity")
+  /**
+   * Read TableMetadata from a JSON string.
+   * <p>
+   * The TableMetadata's metadata file location will be unset.
+   *
+   * @param io a FileIO used by {@link Snapshot} instances
+   * @param json a JSON string of table metadata
+   * @return a TableMetadata object
+   */
+  public static TableMetadata fromJson(FileIO io, String json) {
```

Comment:
For serializing/deserializing with REST catalog, it would be nice to have an option to parse the metadata and then later bind in the FileIO.

Reply:
You can pass `null` for `FileIO` in those cases. It isn't used unless you call methods on `Snapshot` that need to open files, like `addedFiles`. To fix that, I think we want to gracefully handle missing `FileIO` in `Snapshot`, but that's a separate PR.

Reply:
I went to add the Snapshot changes and they're actually smaller than I thought so I'll add them in this PR. I still think it's a good idea to require `FileIO` in the method because we want most people to always pass one.